File: app/services/extraction/pdf_extractor.py
"""
PDF statement extractor using pdfplumber.
Detects Nigerian bank format and extracts transactions.
"""
from __future__ import annotations

import re
import logging
from dataclasses import dataclass, field
from datetime import datetime, date
from pathlib import Path

import pdfplumber

logger = logging.getLogger(__name__)

NIGERIAN_BANKS = [
    "access bank", "zenith bank", "gtbank", "guaranty trust", "first bank",
    "uba", "united bank for africa", "fidelity bank", "sterling bank",
    "ecobank", "union bank", "stanbic ibtc", "keystone bank", "heritage bank",
    "wema bank", "providus bank", "titan trust", "polaris bank",
    "opay", "kuda", "moniepoint", "palmpay", "vfd microfinance",
]

DATE_PATTERNS = [
    re.compile(r"\b(\d{2}[/\-]\d{2}[/\-]\d{4})\b"),
    re.compile(r"\b(\d{2}[/\-]\d{2}[/\-]\d{2})(?:\s+\d{2}:\d{2}(?::\d{2})?)?\b"),
    re.compile(r"\b(\d{4}[/\-]\d{2}[/\-]\d{2})\b"),
    re.compile(r"\b(\d{2}\s+(?:Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)[a-z]*\s+\d{4})\b", re.I),
    re.compile(r"\b((?:Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)[a-z]*\s+\d{1,2},?\s+\d{4})\b", re.I),
]
AMOUNT_RE = re.compile(r"[\d,]+\.\d{2}")
BALANCE_KEYWORDS = re.compile(r"(opening|closing|balance|brought forward|b/f|c/f)", re.I)


@dataclass
class ExtractedTransaction:
    date: date
    description: str
    amount: float
    type: str   # debit | credit
    balance: float | None = None
    raw_date: str = ""
    raw_amount: str = ""
    raw_balance: str = ""


@dataclass
class ExtractionResult:
    transactions: list[ExtractedTransaction] = field(default_factory=list)
    bank_name: str | None = None
    account_number: str | None = None
    period_from: date | None = None
    period_to: date | None = None
    opening_balance: float | None = None
    closing_balance: float | None = None
    warnings: list[str] = field(default_factory=list)


class PdfExtractor:
    def extract(self, file_path: str | Path) -> ExtractionResult:
        result = ExtractionResult()
        try:
            with pdfplumber.open(str(file_path)) as pdf:
                full_text = "\n".join(page.extract_text() or "" for page in pdf.pages)

            result.bank_name = self._detect_bank(full_text)
            result.opening_balance, result.closing_balance = self._extract_balances(full_text)
            result.period_from, result.period_to = self._extract_period(full_text)

            # Try table extraction first (structured PDFs)
            with pdfplumber.open(str(file_path)) as pdf:
                all_rows: list[list[str | None]] = []
                for page in pdf.pages:
                    tables = page.extract_tables()
                    for table in tables:
                        all_rows.extend(table)

            if all_rows:
                txs = self._parse_table_rows(all_rows)
                if txs:
                    result.transactions = txs
                    return result

            # Fallback: line-by-line text parsing
            result.transactions = self._parse_text_lines(full_text)
            if not result.transactions:
                result.warnings.append("Could not extract transactions from this PDF format")

        except Exception as e:
            logger.error(f"PDF extraction error: {e}")
            result.warnings.append(f"Extraction error: {str(e)}")

        return result

    def _detect_bank(self, text: str) -> str | None:
        lower = text[:500].lower()
        for bank in NIGERIAN_BANKS:
            if bank in lower:
                return bank.title()
        return None

    def _extract_balances(self, text: str) -> tuple[float | None, float | None]:
        opening = None
        closing = None
        lines = text.split("\n")
        for line in lines:
            lower = line.lower()
            amounts = AMOUNT_RE.findall(line)
            if not amounts:
                continue
            val = float(amounts[-1].replace(",", ""))
            if "opening" in lower or "brought forward" in lower or "b/f" in lower:
                opening = val
            elif "closing" in lower or "carried forward" in lower or "c/f" in lower:
                closing = val
        return opening, closing

    def _extract_period(self, text: str) -> tuple[date | None, date | None]:
        dates: list[date] = []
        for pattern in DATE_PATTERNS:
            for match in pattern.finditer(text[:800]):
                parsed = self._parse_date(match.group(1))
                if parsed:
                    dates.append(parsed)
        if len(dates) >= 2:
            return min(dates), max(dates)
        return None, None

    def _parse_table_rows(self, rows: list[list[str | None]]) -> list[ExtractedTransaction]:
        txs: list[ExtractedTransaction] = []
        date_col = amount_col = desc_col = debit_col = credit_col = balance_col = -1

        # Find header row
        for i, row in enumerate(rows[:5]):
            cells = [str(c or "").lower() for c in row]
            for j, cell in enumerate(cells):
                if any(k in cell for k in ["date", "value date", "trans date"]):
                    date_col = j
                if any(k in cell for k in ["description", "narration", "details", "particulars"]):
                    desc_col = j
                if any(k in cell for k in ["amount", "debit", "withdrawal"]) and "credit" not in cell:
                    if debit_col == -1:
                        debit_col = j
                if any(k in cell for k in ["credit", "deposit"]):
                    credit_col = j
                if any(k in cell for k in ["balance"]):
                    balance_col = j
            if date_col >= 0 and desc_col >= 0:
                rows = rows[i + 1:]
                break

        if date_col < 0 or desc_col < 0:
            return []

        for row in rows:
            if not row or all(c is None or str(c).strip() == "" for c in row):
                continue
            try:
                raw_date = str(row[date_col] or "").strip()
                parsed_date = self._parse_date(raw_date)
                if not parsed_date:
                    continue

                desc = str(row[desc_col] or "").strip()
                if not desc:
                    continue

                debit_val = credit_val = 0.0
                if debit_col >= 0 and debit_col < len(row):
                    debit_val = self._parse_amount(str(row[debit_col] or ""))
                if credit_col >= 0 and credit_col < len(row):
                    credit_val = self._parse_amount(str(row[credit_col] or ""))

                if amount_col >= 0 and amount_col < len(row) and debit_val == 0 and credit_val == 0:
                    amt = self._parse_amount(str(row[amount_col] or ""))
                    if amt > 0:
                        debit_val = amt

                if debit_val == 0 and credit_val == 0:
                    continue

                balance = None
                if balance_col >= 0 and balance_col < len(row):
                    balance = self._parse_amount(str(row[balance_col] or "")) or None

                tx_type = "credit" if credit_val > debit_val else "debit"
                amount = credit_val if tx_type == "credit" else debit_val

                txs.append(ExtractedTransaction(
                    date=parsed_date, description=desc,
                    amount=amount, type=tx_type,
                    balance=balance, raw_date=raw_date,
                    raw_amount=str(amount), raw_balance=str(balance or ""),
                ))
            except Exception:
                continue

        return txs

    def _parse_text_lines(self, text: str) -> list[ExtractedTransaction]:
        txs: list[ExtractedTransaction] = []
        lines = text.split("\n")

        for line in lines:
            line = line.strip()
            if not line or len(line) < 15:
                continue

            date_match = None
            raw_date = ""
            for pattern in DATE_PATTERNS:
                m = pattern.search(line)
                if m:
                    date_match = self._parse_date(m.group(1))
                    raw_date = m.group(1)
                    break

            if not date_match:
                continue

            amounts = AMOUNT_RE.findall(line)
            if not amounts:
                continue

            amount_vals = [float(a.replace(",", "")) for a in amounts]
            balance = amount_vals[-1] if len(amount_vals) >= 2 else None
            amount = amount_vals[-2] if len(amount_vals) >= 2 else amount_vals[0]

            # Remove date, amounts from description
            desc = line
            for pattern in DATE_PATTERNS:
                desc = pattern.sub("", desc)
            desc = AMOUNT_RE.sub("", desc).strip(" |-/,")

            if not desc or amount <= 0:
                continue

            lower = line.lower()
            tx_type = "credit" if any(k in lower for k in ["cr", "credit", "deposit", "received"]) else "debit"

            txs.append(ExtractedTransaction(
                date=date_match, description=desc[:200],
                amount=amount, type=tx_type,
                balance=balance, raw_date=raw_date,
                raw_amount=str(amount), raw_balance=str(balance or ""),
            ))

        return txs

    def _parse_date(self, raw: str) -> date | None:
        formats = [
            "%d/%m/%Y", "%d-%m-%Y", "%Y/%m/%d", "%Y-%m-%d",
            "%d/%m/%y", "%d-%m-%y",
            "%d %b %Y", "%d %B %Y", "%b %d, %Y", "%B %d, %Y",
            "%d%b%Y", "%d/%b/%Y", "%d-%b-%Y", "%d-%b-%y",
        ]
        raw = raw.strip()
        for fmt in formats:
            try:
                return datetime.strptime(raw, fmt).date()
            except ValueError:
                continue
        return None

    def _parse_amount(self, raw: str) -> float:
        cleaned = re.sub(r"[^\d.]", "", raw.replace(",", ""))
        try:
            return float(cleaned)
        except ValueError:
            return 0.0


pdf_extractor = PdfExtractor()

File: src/services/extraction/pdfExtractor.ts
import pdfParse from "pdf-parse";
import type { ExtractionResult, RawTransaction } from "../../types";
import { parseAmount, parseDate } from "../utils/parseHelpers";

// ─── PDF Text Table Parser ────────────────────────────────────────────────────

export async function extractFromPDF(buffer: Buffer): Promise<ExtractionResult> {
  const warnings: string[] = [];
  const errors: string[] = [];
  const transactions: RawTransaction[] = [];

  let text = "";

  try {
    const data = await pdfParse(buffer, {
      max: 0, // parse all pages
    });
    text = data.text;
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    errors.push(`PDF extraction failed: ${message}`);
    return { transactions, warnings, errors };
  }

  if (!text || text.trim().length < 50) {
    errors.push("PDF appears to be scanned or empty. OCR support coming soon.");
    return { transactions, warnings, errors };
  }

  const lines = text
    .split("\n")
    .map((l) => l.trim())
    .filter((l) => l.length > 0);

  // Try to detect statement structure
  const structureResult = detectStatementStructure(lines, warnings);

  if (structureResult.transactions.length === 0) {
    warnings.push("Could not parse transactions using standard format. Attempting line-by-line parsing.");
    const fallback = parseLineByLine(lines, warnings);
    transactions.push(...fallback);
  } else {
    transactions.push(...structureResult.transactions);
  }

  return {
    transactions,
    openingBalance: structureResult.openingBalance,
    closingBalance: structureResult.closingBalance,
    periodStart: structureResult.periodStart,
    periodEnd: structureResult.periodEnd,
    bankName: structureResult.bankName,
    warnings,
    errors,
  };
}

interface StructureResult {
  transactions: RawTransaction[];
  openingBalance?: number;
  closingBalance?: number;
  periodStart?: Date;
  periodEnd?: Date;
  bankName?: string;
}

function detectStatementStructure(lines: string[], warnings: string[]): StructureResult {
  const transactions: RawTransaction[] = [];
  let openingBalance: number | undefined;
  let closingBalance: number | undefined;
  let periodStart: Date | undefined;
  let periodEnd: Date | undefined;
  let bankName: string | undefined;

  // Detect known Nigerian bank headers
  const BANK_PATTERNS: [RegExp, string][] = [
    [/GUARANTY\s*TRUST|GTBank|GTB\b/i, "GTBank"],
    [/ZENITH\s*BANK/i, "Zenith Bank"],
    [/ACCESS\s*BANK/i, "Access Bank"],
    [/UNITED\s*BANK.*AFRICA|UBA\b/i, "UBA"],
    [/FIRST\s*BANK/i, "First Bank"],
    [/STANBIC\s*IBTC/i, "Stanbic IBTC"],
    [/FIDELITY\s*BANK/i, "Fidelity Bank"],
    [/POLARIS\s*BANK/i, "Polaris Bank"],
    [/STERLING\s*BANK/i, "Sterling Bank"],
    [/ECOBANK/i, "Ecobank"],
    [/UNION\s*BANK/i, "Union Bank"],
    [/WEMA\s*BANK/i, "Wema Bank"],
    [/KUDA\s*BANK/i, "Kuda Bank"],
    [/OPAY/i, "OPay"],
    [/PALMPAY/i, "PalmPay"],
  ];

  for (const line of lines.slice(0, 20)) {
    for (const [pattern, name] of BANK_PATTERNS) {
      if (pattern.test(line)) {
        bankName = name;
        break;
      }
    }
    if (bankName) break;
  }

  // Look for opening/closing balance
  for (const line of lines) {
    const openMatch = line.match(/opening\s*balance[:\s]+([₦\d,\.]+)/i);
    if (openMatch) openingBalance = parseAmount(openMatch[1]) ?? undefined;

    const closeMatch = line.match(/closing\s*balance[:\s]+([₦\d,\.]+)/i);
    if (closeMatch) closingBalance = parseAmount(closeMatch[1]) ?? undefined;

    const periodMatch = line.match(/(\d{1,2}[\/\-]\d{1,2}[\/\-]\d{2,4})\s+to\s+(\d{1,2}[\/\-]\d{1,2}[\/\-]\d{2,4})/i);
    if (periodMatch) {
      periodStart = parseDate(periodMatch[1]) ?? undefined;
      periodEnd = parseDate(periodMatch[2]) ?? undefined;
    }
  }

  // Try to find and parse transaction table
  // Pattern: DATE | DESCRIPTION | DEBIT | CREDIT | BALANCE (or variations)
  const transactionLines = extractTransactionLines(lines, warnings);
  transactions.push(...transactionLines);

  return { transactions, openingBalance, closingBalance, periodStart, periodEnd, bankName };
}

function extractTransactionLines(lines: string[], warnings: string[]): RawTransaction[] {
  const transactions: RawTransaction[] = [];

  // Nigerian bank statement date patterns
  const DATE_REGEX = /\b(\d{1,2}[\/\-]\d{1,2}[\/\-]\d{2,4}|\d{2}[A-Za-z]{3}\d{2,4}|\d{4}-\d{2}-\d{2})\b/;
  const AMOUNT_REGEX = /[₦]?\s*[\d,]+\.\d{2}/g;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];

    // Check if line starts with a date
    const dateMatch = line.match(DATE_REGEX);
    if (!dateMatch) continue;

    const date = parseDate(dateMatch[1]);
    if (!date || date.getFullYear() < 2000) continue;

    // Extract all amounts from this line
    const amounts: number[] = [];
    let match: RegExpExecArray | null;
    const amountRegex = /[₦]?\s*([\d,]+\.\d{2})/g;
    while ((match = amountRegex.exec(line)) !== null) {
      const amt = parseAmount(match[1]);
      if (amt !== undefined && amt > 0) amounts.push(amt);
    }

    if (amounts.length === 0) continue;

    // Get description - the non-numeric, non-date part
    const description = line
      .replace(DATE_REGEX, "")
      .replace(/[₦]?\s*[\d,]+\.\d{2}/g, "")
      .replace(/\s+/g, " ")
      .trim();

    if (description.length < 2) continue;

    // Determine debit/credit based on position (last amount is usually balance)
    let debit: number | undefined;
    let credit: number | undefined;
    let balance: number | undefined;

    if (amounts.length === 1) {
      // Could be either, default to debit
      debit = amounts[0];
    } else if (amounts.length === 2) {
      // First is transaction amount, second is balance
      debit = amounts[0];
      balance = amounts[1];
    } else if (amounts.length >= 3) {
      // Debit, credit, balance pattern
      if (amounts[0] > 0 && amounts[1] === 0) {
        debit = amounts[0];
      } else if (amounts[0] === 0 && amounts[1] > 0) {
        credit = amounts[1];
      } else {
        debit = amounts[0];
        credit = amounts[1];
      }
      balance = amounts[amounts.length - 1];
    }

    const type: "debit" | "credit" | "transfer" =
      credit !== undefined && (debit === undefined || debit === 0) ? "credit" : "debit";

    transactions.push({
      date: dateMatch[1],
      description,
      debit: type === "debit" ? (debit || amounts[0]) : undefined,
      credit: type === "credit" ? credit : undefined,
      amount: debit || credit || amounts[0],
      balance,
      type,
    });
  }

  return transactions;
}

function parseLineByLine(lines: string[], warnings: string[]): RawTransaction[] {
  const transactions: RawTransaction[] = [];
  warnings.push("Using fallback line-by-line parser");

  for (const line of lines) {
    // Very basic: look for a date anywhere in the line
    const dateMatch = line.match(/\b(\d{1,2}[\/\-]\d{1,2}[\/\-]\d{2,4})\b/);
    const amountMatch = line.match(/[₦]?\s*([\d,]+\.\d{2})/);

    if (!dateMatch || !amountMatch) continue;

    const date = parseDate(dateMatch[1]);
    const amount = parseAmount(amountMatch[1]);

    if (!date || amount === undefined || amount <= 0) continue;

    const description = line
      .replace(/\b\d{1,2}[\/\-]\d{1,2}[\/\-]\d{2,4}\b/, "")
      .replace(/[₦]?\s*[\d,]+\.\d{2}/g, "")
      .replace(/\s+/g, " ")
      .trim();

    if (!description) continue;

    transactions.push({
      date: dateMatch[1],
      description,
      amount,
      type: "debit",
    });
  }

  return transactions;
}
